substratevm/src/com.oracle.svm.core.jdk9/src/com/oracle/svm/core/jdk/Target_java_lang_StackWalker.java

```diff
@@ -25,19 +25,367 @@
+                    FrameInfoQueryResult frameInfo = frame.getFrameInfo();
+                    String className = frameInfo.getSourceClassName();
+                    if (includeHidden || ! StackWalker_Util.isHidden(className) || includeReflect || ! StackWalker_Util.isReflect(className)) {
+                        return new StackFrameImpl(frameInfo).getDeclaringClass();
+                    }
+                }
+            } else {
+                CodeInfoQueryResult codeInfo = CodeInfoTable.lookupCodeInfoQueryResult(ip);
+                for (FrameInfoQueryResult frameInfo = codeInfo.getFrameInfo(); frameInfo != null; frameInfo = frameInfo.getCaller()) {
+                    String className = frameInfo.getSourceClassName();
+                    if (includeHidden || ! StackWalker_Util.isHidden(className) || includeReflect || ! StackWalker_Util.isReflect(className)) {
+                        return new StackFrameImpl(frameInfo).getDeclaringClass();
+                    }
+                }
+            }
+        } while (JavaStackWalker.continueWalk(walk));
+        throw new IllegalStateException("No calling frame");
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance() {
+        return StackWalkerHolder.PLAIN;
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance(StackWalker.Option option) {
+        if (option == StackWalker.Option.SHOW_HIDDEN_FRAMES) {
+            return StackWalkerHolder.SHF;
+        } else if (option == StackWalker.Option.SHOW_REFLECT_FRAMES) {
+            return StackWalkerHolder.SRF;
+        } else {
+            return StackWalkerHolder.PLAIN;
+        }
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance(Set<StackWalker.Option> option) {
+        if (option.contains(StackWalker.Option.SHOW_HIDDEN_FRAMES)) {
+            return StackWalkerHolder.SHF;
+        } else if (option.contains(StackWalker.Option.SHOW_REFLECT_FRAMES)) {
+            return StackWalkerHolder.SRF;
+        } else {
+            return StackWalkerHolder.PLAIN;
+        }
+    }
+
+    @Substitute
+    public static Target_java_lang_StackWalker getInstance(Set<StackWalker.Option> option, int estDepth) {
+        return getInstance(option);
+    }
+}
+
+final class StackWalker_Util {
+    private StackWalker_Util() {}
+
+    static boolean isReflect(final String nextClassName) {
+        return Method.class.getName().equals(nextClassName) || Constructor.class.getName().equals(nextClassName);
+        // TODO: MethodAccessor.class.isAssignableFrom(c)
+        // TODO: ConstructorAccessor.class.isAssignableFrom(c)
+        // TODO: c.getName().startsWith("java.lang.invoke.LambdaForm")
+    }
+
+    static boolean isHidden(final String nextClassName) {
+        return ImplicitExceptions.class.getName().equals(nextClassName);
+    }
+}
+
+final class StackIterator implements Iterator<StackFrame> {
```

Comment:
Implementing this as an `Iterator` is overly complicated, because of the `hasNext` - `next` separation. I converted it to a `Spliterator` and now the whole implementation is in a single method.

Reply:
Sounds good.  I only went with `Iterator` as it was something I'm very familiar with.